src/compiler/moduleSpecifiers.ts

```diff
@@ -106,9 +106,19 @@ namespace ts.moduleSpecifiers {
         if (!moduleSourceFile) {
             return [];
         }
-        const modulePaths = getAllModulePaths(importingSourceFile.path, moduleSourceFile.originalFileName, host);
-        const preferences = getPreferences(userPreferences, compilerOptions, importingSourceFile);
 
+        const cache = host.getModuleSpecifierCache?.();
+        const cached = cache?.get(importingSourceFile.path, moduleSourceFile.path);
+        let modulePaths;
+        if (typeof cached === "object") {
+            if (cached.moduleSpecifiers) return cached.moduleSpecifiers;
+            modulePaths = cached.modulePaths;
+        }
+        else {
+            modulePaths = getAllModulePathsWorker(importingSourceFile.path, moduleSourceFile.originalFileName, host);

```

Comment:
What if cache said file was not importable.. (false value) shouldnt you be skipping the work in that case ?

Reply:
We currently never call into this function if we already have a `false` result, because we check `isImportableFile` first, which is a cheaper subset of this work that caches a boolean. It’s theoretically possible that a different caller would want to get these module paths even if auto-imports determined that none of them were suitable for import (e.g., the checker calls this function and always expects to get a non-empty result).

I’m not really convinced that storing the boolean in the cache is really adding much to the performance advantage and this line of questioning makes it clear that it’s a little confusing to think about. I plan to investigate removing it later to simplify things a bit.

Reply:
It does look like the boolean (probably just `false`) is somewhat valuable (takes 20ms off a 100ms response) for pnpm where there are lots of modules in paths that can't be imported. But I’m going to add it as a separate property or store it in a separate map instead of overloading the cached value since it’s actually somewhat orthogonal to the module paths computation.